### PR TITLE
MULE-18873: Scatter-Gather routes are dropping events during Flow-Ref processing

### PR DESCRIPTION
**Pull request by IvanAndresFritzler**

### CODE REVIEW

File: core-tests/src/test/java/org/mule/runtime/core/internal/construct/DefaultFlowTestCase.java [hunk 1]
```diff
@@ -503,7 +520,30 @@ public void doNotExecuteOnErrorContinueDefinedOutsideTheFlow() throws MuleExcept
...
+    Flux<CoreEvent> flowProcessing = Flux
+        .from(MessageProcessors.processWithChildContext(testEvent, flow, Optional.of(from(FLOW_NAME))));
+    Future validation = executor.submit(() -> StepVerifier.create(flowProcessing)
+        .expectNext(testEvent)
+        .expectComplete()
+        .verifyThenAssertThat()
+        .hasNotDroppedElements());
+    validation.get();
+  }
+
```
Comment (eze210) [inline core-tests/src/test/java/org/mule/runtime/core/internal/construct/DefaultFlowTestCase.java:543]: remove extra blanks

File: core-tests/src/test/java/org/mule/runtime/core/internal/construct/DefaultFlowTestCase.java [hunk 2]
```diff
@@ -524,4 +564,18 @@ public CoreEvent process(CoreEvent event) throws MuleException {
           });
     }
   }
+
+  public static class BlockMessageProcessor extends AbstractComponent implements Processor {
+
+    @Override
+    public CoreEvent process(CoreEvent event) throws MuleException {
+      return sleepFor(event, muleContext.getConfiguration().getShutdownTimeout() * 2);
```
Comment (eze210) [inline core-tests/src/test/java/org/mule/runtime/core/internal/construct/DefaultFlowTestCase.java:572]: Isn't it possible to use a Latch?

File: core-tests/src/test/java/org/mule/runtime/core/internal/construct/DefaultFlowTestCase.java [hunk 3]
```diff
@@ -503,7 +520,30 @@ public void doNotExecuteOnErrorContinueDefinedOutsideTheFlow() throws MuleExcept
...
+      throws MuleException, ExecutionException, InterruptedException {
+    CoreEvent testEvent = testEvent();
+    flow = (DefaultFlow) Flow.builder(FLOW_NAME, muleContext)
+        .source(flow.getSource())
+        .processors(singletonList(new BlockMessageProcessor()))
+        .processingStrategyFactory(new ProactorStreamEmitterProcessingStrategyFactory())
+        .build();
+    startFlow();
+    Flux<CoreEvent> flowProcessing = Flux
+        .from(MessageProcessors.processWithChildContext(testEvent, flow, Optional.of(from(FLOW_NAME))));
```
Comment (eze210) [inline core-tests/src/test/java/org/mule/runtime/core/internal/construct/DefaultFlowTestCase.java:535]: static import

File: core-tests/src/test/java/org/mule/runtime/core/internal/construct/DefaultFlowTestCase.java [hunk 4]
```diff
@@ -503,7 +520,30 @@ public void doNotExecuteOnErrorContinueDefinedOutsideTheFlow() throws MuleExcept
...
+    Future validation = executor.submit(() -> StepVerifier.create(flowProcessing)
```
Comment (eze210) [inline core-tests/src/test/java/org/mule/runtime/core/internal/construct/DefaultFlowTestCase.java:536]: Is the executor needed? What do you think about directly this?
```
StepVerifier.create(flowProcessing)
        .expectNext(testEvent)
        .expectComplete()
        .verifyThenAssertThat()
        .hasNotDroppedElements();
```

**Review by eze210 — CHANGES_REQUESTED** [review]

File: core-tests/src/test/java/org/mule/runtime/core/internal/construct/DefaultFlowTestCase.java (hunk 2 above)
Comment (IvanAndresFritzler) [reply core-tests/src/test/java/org/mule/runtime/core/internal/construct/DefaultFlowTestCase.java:572]: Discussed offline, but short answer is that a latch has no use for this test case (we just couldn't know when such latch should be released)

File: core-tests/src/test/java/org/mule/runtime/core/internal/construct/DefaultFlowTestCase.java (hunk 4 above)
Comment (IvanAndresFritzler) [reply core-tests/src/test/java/org/mule/runtime/core/internal/construct/DefaultFlowTestCase.java:536]: No needed at all (was a residue of trying to see if a latch could be used)

**Review by eze210 — APPROVED** [review]

**Pull request merged**